pkg/kubelet/images/pullmanager/fs_pullrecords.go

```diff
@@ -0,0 +1,275 @@
+	intentBytes, err := os.ReadFile(intentRecordPath)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	intent, err := decodeIntent(f.decoder, intentBytes)
+	if err != nil {
+		return false, err
+	}
+
+	return intent.Image == image, nil
+}
+
+func (f *fsPullRecordsAccessor) DeleteImagePullIntent(image string) error {
+	intentRecordPath := filepath.Join(f.pullingDir, cacheFilename(image))
+	if err := os.Remove(intentRecordPath); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func (f *fsPullRecordsAccessor) GetImagePulledRecord(imageRef string) (*kubeletconfiginternal.ImagePulledRecord, bool, error) {
+	recordBytes, err := os.ReadFile(filepath.Join(f.pulledDir, cacheFilename(imageRef)))
+	if os.IsNotExist(err) {
+		return nil, false, nil
+	} else if err != nil {
+		return nil, false, err
+	}
+
+	pulledRecord, err := decodePulledRecord(f.decoder, recordBytes)
+	return pulledRecord, true, err
+}
+
+func (f *fsPullRecordsAccessor) ListImagePulledRecords() ([]*kubeletconfiginternal.ImagePulledRecord, error) {
+	var pullRecords []*kubeletconfiginternal.ImagePulledRecord
+	err := processDirFiles(f.pulledDir,
+		func(filePath string, fileContent []byte) error {
+			pullRecord, err := decodePulledRecord(f.decoder, fileContent)
+			if err != nil {
+				klog.V(5).InfoS("failed to deserialize, skipping file", "filePath", filePath, "error", err)
+				return nil
+			}
+			pullRecords = append(pullRecords, pullRecord)
+			return nil
+		})
+
+	return pullRecords, err
+}
+
+func (f *fsPullRecordsAccessor) WriteImagePulledRecord(pulledRecord *kubeletconfiginternal.ImagePulledRecord) error {
+	recordBytes := bytes.NewBuffer([]byte{})
+	if err := f.encoder.Encode(pulledRecord, recordBytes); err != nil {
+		return fmt.Errorf("failed to serialize ImagePulledRecord: %w", err)
+	}
+
+	return writeFile(f.pulledDir, cacheFilename(pulledRecord.ImageRef), recordBytes.Bytes())
+}
+
+func (f *fsPullRecordsAccessor) DeleteImagePulledRecord(imageRef string) error {
+	err := os.Remove(filepath.Join(f.pulledDir, cacheFilename(imageRef)))
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func cacheFilename(image string) string {
+	return fmt.Sprintf("sha256-%x", sha256.Sum256([]byte(image)))
+}
+
+// writeFile writes `content` to the file with name `filename` in directory `dir`.
+// It assures write atomicity by creating a temporary file first and only after
+// a successful write, it move the temp file in place of the target.
+func writeFile(dir, filename string, content []byte) error {
+	// create target folder if it does not exists yet
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return fmt.Errorf("failed to create directory %q: %w", dir, err)
+	}
+
+	targetPath := filepath.Join(dir, filename)
+	tmpPath := targetPath + ".tmp"
+	if err := os.WriteFile(tmpPath, content, 0600); err != nil {
+		return fmt.Errorf("failed to create temporary file %q: %w", tmpPath, err)
```

Comment:
do a best-effort delete of the tmp file here in case it was partially written (ignore error)